File: web/src/services/zquant/roles.ts
// @ts-ignore
// Copyright 2025 ZQuant Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: kevin
// Contact:
//     - Email: kevin@vip.qq.com
//     - Wechat: zquant2025
//     - Issues: https://github.com/yoyoung/zquant/issues
//     - Documentation: https://github.com/yoyoung/zquant/blob/main/README.md
//     - Repository: https://github.com/yoyoung/zquant

/* eslint-disable */
import { request } from '@umijs/max';

/**
 * 查询角色列表
 * POST /api/v1/roles
 */
export async function getRoles(params?: {
  skip?: number;
  limit?: number;
  order_by?: string;
  order?: 'asc' | 'desc';
}) {
  return request<ZQuant.PageResponse<ZQuant.RoleResponse>>('/api/v1/roles', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: params || {},
  });
}

/**
 * 查询角色详情
 * POST /api/v1/roles/get
 */
export async function getRole(roleId: number) {
  return request<ZQuant.RoleWithPermissions>('/api/v1/roles/get', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { role_id: roleId },
  });
}

/**
 * 创建角色
 * POST /api/v1/roles
 */
export async function createRole(body: ZQuant.RoleCreate) {
  return request<ZQuant.RoleResponse>('/api/v1/roles', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
  });
}

/**
 * 更新角色
 * POST /api/v1/roles/update
 */
export async function updateRole(roleId: number, body: ZQuant.RoleUpdate) {
  return request<ZQuant.RoleResponse>('/api/v1/roles/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, role_id: roleId },
  });
}

/**
 * 删除角色
 * POST /api/v1/roles/delete
 */
export async function deleteRole(roleId: number) {
  return request<{ message: string }>('/api/v1/roles/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { role_id: roleId },
  });
}

/**
 * 查询角色的权限列表
 * POST /api/v1/roles/permissions/list
 */
export async function getRolePermissions(roleId: number) {
  return request<ZQuant.PermissionResponse[]>('/api/v1/roles/permissions/list', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { role_id: roleId },
  });
}

/**
 * 为角色分配权限
 * POST /api/v1/roles/permissions/assign
 */
export async function assignPermissions(roleId: number, body: ZQuant.AssignPermissionsRequest) {
  return request<ZQuant.RoleResponse>('/api/v1/roles/permissions/assign', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { ...body, role_id: roleId },
  });
}

/**
 * 为角色添加单个权限
 * POST /api/v1/roles/permissions/add
 */
export async function addPermission(roleId: number, permissionId: number) {
  return request<{ message: string }>('/api/v1/roles/permissions/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { role_id: roleId, permission_id: permissionId },
  });
}

/**
 * 移除角色的单个权限
 * POST /api/v1/roles/permissions/remove
 */
export async function removePermission(roleId: number, permissionId: number) {
  return request<{ message: string }>('/api/v1/roles/permissions/remove', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: { role_id: roleId, permission_id: permissionId },
  });
}

File: zquant/api/v1/roles.py
# Copyright 2025 ZQuant Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Author: kevin
# Contact:
#     - Email: kevin@vip.qq.com
#     - Wechat: zquant2025
#     - Issues: https://github.com/yoyoung/zquant/issues
#     - Documentation: https://github.com/yoyoung/zquant/blob/main/README.md
#     - Repository: https://github.com/yoyoung/zquant

"""
角色管理API
"""

from fastapi import APIRouter, Depends, HTTPException, status
from loguru import logger
from sqlalchemy.orm import Session

from zquant.api.deps import get_current_active_user
from zquant.core.exceptions import NotFoundError, ValidationError
from zquant.core.permissions import check_permission
from zquant.database import get_db
from zquant.models.user import User
from zquant.schemas.common import QueryRequest
from zquant.schemas.user import (
    AssignPermissionsRequest,
    PageResponse,
    PermissionResponse,
    RoleCreate,
    RoleDeleteRequest,
    RoleGetRequest,
    RoleListRequest,
    RolePermissionAddRequest,
    RolePermissionRemoveRequest,
    RolePermissionsListRequest,
    RoleResponse,
    RoleUpdate,
    RoleWithPermissions,
)
from zquant.services.role import RoleService

router = APIRouter()


@router.post("", response_model=PageResponse, summary="查询角色列表")
@check_permission("role", "read")
def get_roles(
    request: RoleListRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """查询角色列表（分页、排序）"""
    roles = RoleService.get_all_roles(
        db, skip=request.skip, limit=request.limit, order_by=request.order_by, order=request.order
    )
    total = RoleService.count_roles(db)
    return PageResponse(
        items=[RoleResponse.model_validate(r) for r in roles], total=total, skip=request.skip, limit=request.limit
    )


@router.post("/get", response_model=RoleWithPermissions, summary="查询角色详情")
@check_permission("role", "read")
def get_role(request: RoleGetRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)):
    """根据ID查询角色详情（包含权限列表）"""
    role = RoleService.get_role_by_id(db, request.role_id)
    if not role:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"角色ID {request.role_id} 不存在")

    permissions = RoleService.get_role_permissions(db, request.role_id)
    # 构造响应对象
    role_dict = {
        "id": role.id,
        "name": role.name,
        "description": role.description,
        "created_time": role.created_time,
        "permissions": [PermissionResponse.model_validate(p) for p in permissions],
    }
    return RoleWithPermissions.model_validate(role_dict)


@router.post("", response_model=RoleResponse, status_code=status.HTTP_201_CREATED, summary="创建角色")
@check_permission("role", "create")
def create_role(
    role_data: RoleCreate, db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)
):
    """创建角色（需要role:create权限）"""
    try:
        role = RoleService.create_role(db, role_data)
        return role
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/update", response_model=RoleResponse, summary="更新角色")
@check_permission("role", "update")
def update_role(
    role_data: RoleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """更新角色信息（需要role:update权限）"""
    try:
        role = RoleService.update_role(db, role_data.role_id, role_data)
        return role
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/delete", summary="删除角色")
@check_permission("role", "delete")
def delete_role(request: RoleDeleteRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)):
    """删除角色（需要role:delete权限）"""
    try:
        RoleService.delete_role(db, request.role_id)
        return {"message": "角色已删除"}
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/permissions/list", response_model=list[PermissionResponse], summary="查询角色的权限列表")
@check_permission("role", "read")
def get_role_permissions(
    request: RolePermissionsListRequest, db: Session = Depends(get_db), current_user: User = Depends(get_current_active_user)
):
    """查询角色的权限列表"""
    try:
        permissions = RoleService.get_role_permissions(db, request.role_id)
        return [PermissionResponse.model_validate(p) for p in permissions]
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.post("/permissions/assign", response_model=RoleResponse, summary="为角色分配权限")
@check_permission("role", "update")
def assign_permissions(
    request: AssignPermissionsRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """为角色分配权限（需要role:update权限）"""
    try:
        role = RoleService.assign_permissions(db, request.role_id, request.permission_ids)
        return role
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/permissions/add", summary="为角色添加权限")
@check_permission("role", "update")
def add_permission(
    request: RolePermissionAddRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """为角色添加单个权限（需要role:update权限）"""
    try:
        RoleService.add_permission(db, request.role_id, request.permission_id)
        return {"message": "权限已添加"}
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.post("/permissions/remove", summary="移除角色的权限")
@check_permission("role", "update")
def remove_permission(
    request: RolePermissionRemoveRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """移除角色的单个权限（需要role:update权限）"""
    try:
        RoleService.remove_permission(db, request.role_id, request.permission_id)
        return {"message": "权限已移除"}
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
